### PR TITLE
Resolve #2017: Make constructors public

### LINKED ISSUE #2017: Index predicates: Make constructors public
In order to make use of the new IndexPredicate, constructors need to be made public

### PR DESCRIPTION
**Pull request by ohadzeliger**

### CODE REVIEW

**Review by hatyo — APPROVED** [review]
Nice work @ohadzeliger, it looks good to me, but I would ask @MMcM for a second review if that's ok.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaData.java [hunk 1]
```diff
@@ -99,6 +99,25 @@ public class RecordMetaData implements RecordMetaDataProvider {
...
+                Collections.unmodifiableMap(orig.recordTypeKeyToSyntheticTypeMap),
+                Collections.unmodifiableMap(orig.indexes),
+                Collections.unmodifiableMap(orig.universalIndexes),
+                Collections.unmodifiableList(orig.formerIndexes),
+                orig.splitLongRecords,
+                orig.storeRecordVersions,
+                orig.version,
+                orig.subspaceKeyCounter,
+                orig.usesSubspaceKeyCounter,
+                orig.recordCountKey, // TODO: Do we need to protect from modification?
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaData.java:117]: `KeyExpression`s are immutable.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaData.java [hunk 2]
```diff
@@ -99,6 +99,25 @@ public class RecordMetaData implements RecordMetaDataProvider {
             RecordMetaDataProto.getDescriptor(), RecordMetaDataOptionsProto.getDescriptor(), TupleFieldsProto.getDescriptor()
     };
 
+    protected RecordMetaData(@Nonnull RecordMetaData orig) {
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaData.java:102]: This doesn't seem to be used in this branch, but I wonder whether a `toBuilder` / `asBuilder` would be more idiomatic.

**Review by MMcM — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaData.java (hunk 2 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaData.java:102]: This gets used in the consumer, where a copy-constructor is used to create an isolated instance from the original.
I thought of using a `toBuilder` pattern but ran into difficulties as the building of `RecordMetaDataBuilder` from `RecordMetadaData` is not trivial: For example, the reverse engineering of a a `RecordTypeBuilder` and the Indexes seemed too involved. I would be happy to do so if there is a path though.

Comment (MMcM) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaData.java:102]: I suppose it can wait, then.

**Review by MMcM — APPROVED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaData.java (hunk 1 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaData.java]: Sure. Removed TODO

**Pull request merged**